buildpack/files.go

```diff
@@ -0,0 +1,202 @@
+		return r.Version != ""
+	}
+	return false
+}
+
+func (r *Require) hasInconsistentVersions() bool {
+	if version, ok := r.Metadata["version"]; ok {
+		return r.Version != "" && r.Version != version
+	}
+	return false
+}
+
+func (r *Require) hasTopLevelVersions() bool {
+	return r.Version != ""
+}
+
+type Label struct {
+	Key   string `toml:"key"`
+	Value string `toml:"value"`
+}
+
+// build.toml
+
+type BuildTOML struct {
+	BOM   []BOMEntry `toml:"bom"`
+	Unmet []Unmet    `toml:"unmet"`
+}
+
+type Unmet struct {
+	Name string `toml:"name"`
+}
+
+// store.toml
+
+type StoreTOML struct {
+	Data map[string]interface{} `json:"metadata" toml:"metadata"`
+}
+
+// build plan
+
+type BuildPlan struct {
+	PlanSections
+	Or planSectionsList `toml:"or"`
+}
+
+func (p *PlanSections) hasInconsistentVersions() bool {
+	for _, req := range p.Requires {
+		if req.hasInconsistentVersions() {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *PlanSections) hasDoublySpecifiedVersions() bool {
+	for _, req := range p.Requires {
+		if req.hasDoublySpecifiedVersions() {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *PlanSections) hasTopLevelVersions() bool {
+	for _, req := range p.Requires {
+		if req.hasTopLevelVersions() {
+			return true
+		}
+	}
+	return false
+}
+
+type planSectionsList []PlanSections
+
+func (p *planSectionsList) hasInconsistentVersions() bool {
+	for _, planSection := range *p {
+		if planSection.hasInconsistentVersions() {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *planSectionsList) hasDoublySpecifiedVersions() bool {
+	for _, planSection := range *p {
+		if planSection.hasDoublySpecifiedVersions() {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *planSectionsList) hasTopLevelVersions() bool {
+	for _, planSection := range *p {
+		if planSection.hasTopLevelVersions() {
+			return true
+		}
+	}
+	return false
+}
+
+type PlanSections struct {
+	Requires []Require `toml:"requires"`
+	Provides []Provide `toml:"provides"`
+}
+
+type Provide struct {
+	Name string `toml:"name"`
+}
+
+// buildpack plan
+
+type Plan struct {
+	Entries []Require `toml:"entries"`
+}
+
+func (p Plan) filter(unmet []Unmet) Plan {
+	var out []Require
+	for _, entry := range p.Entries {
+		if !containsName(unmet, entry.Name) {
+			out = append(out, entry)
+		}
+	}
+	return Plan{Entries: out}
+}
+
+func (p Plan) toBOM() []BOMEntry {
+	var bom []BOMEntry
+	for _, entry := range p.Entries {
+		bom = append(bom, BOMEntry{Require: entry})
+	}
+	return bom
+}
+
+func containsName(unmet []Unmet, name string) bool {
+	for _, u := range unmet {
+		if u.Name == name {
+			return true
+		}
+	}
+	return false
+}
+
+// layer content metadata
+
+type LayerMetadataFile struct {
+	Data   interface{} `json:"data" toml:"metadata"`
+	Build  bool        `json:"build" toml:"build"`
+	Launch bool        `json:"launch" toml:"launch"`
+	Cache  bool        `json:"cache" toml:"cache"`
+}
+
+// buildpack.toml - in buildpack/descriptor.go
```

Comment:
I am +1 on breaking out into different files.

Reply:
I like having descriptor as a separate file because at a glance you can see that it's everything to do with buildpack.toml. But I don't think one struct per file is idiomatic in go... plus, buildpack/files.go and platform/files.go has kind of a nice parity. Alternatively, we could move each struct to be near the function where it's used - but where two phases use the same struct it's kind of hard to know where to look for it. WDYT?

Reply:
I'm ok with keeping it as is.

Reply:
Let's keep it in files.go but we can always break it out later.